File: hackerrank.com/problem-solving/divisible_sum_pairs.js
'use strict'

// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem

function divisibleSumPairs(n, k, ar) {
  let res = new Array(k);
  res.fill(0,0,k);
  ar.reduce((acc, i) => { res[i % k]++; return res;}, res);
  let count = 0;
  for (let i = 1; i < Math.floor((k - 1) / 2) + 1; i++) {
    count += res[i] * res[k-i];
  }
  let selfSum = k % 2 === 0 ? [0, k / 2] : [0];
  count = selfSum.reduce((acc, i) => { acc += res[i] * (res[i] - 1) / 2; return acc; }, count);
  return count;
}


function main(inputData) {
  const lines = inputData.split("\n");
  const [n, k] = lines[0].split(' ').map(x => parseInt(x));
  const ar = lines[1].split(' ').map(x => parseInt(x));
  console.log(divisibleSumPairs(n, k, ar))
}

process.stdin.resume();
process.stdin.setEncoding("ascii");
let _input = '';
process.stdin.on("data", input => {
  _input += input;
});

process.stdin.on("end", () => {
 main(_input);
});

File: hackerrank.com/problem-solving/divisible_sum_pairs.scala
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem

import scala.io.StdIn._


object Solution {
  

  def divisibleSumPairs(n: Int, k: Int, ar: Array[Int]): Int = {
    var res: Array[Int] = Array.fill(k){0}
    for (i <- ar) {
      res(i%k) += 1
    }
    var acc = 0
    for (i <- 1 to Math.floor((k - 1) / 2).toInt) {
      acc += res(i) * res(k - i)
    }
    val selfItems = if (k % 2 == 0) Array(0, k / 2) else Array(0)
    for (i <- selfItems) {
      acc += res(i) * (res(i) - 1) / 2
    }
    acc
  }

  def main(args: Array[String]) {
    val nk = readLine.split(" ").map(_.trim.toInt)
    val ar = readLine.split(" ").map(_.trim.toInt)
    println(divisibleSumPairs(nk(0), nk(1), ar))
  }
}